pkg/detectors/bulksms/bulksms.go

```diff
@@ -38,46 +37,66 @@ func (s Scanner) Keywords() []string {
 func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
 	dataStr := string(data)
 
-	matches := keyPat.FindAllStringSubmatch(dataStr, -1)
 	idMatches := idPat.FindAllStringSubmatch(dataStr, -1)
+	keyMatches := keyPat.FindAllStringSubmatch(dataStr, -1)
+
+	verifiedKeys := make(map[string]bool)
+	verifiedIDs := make(map[string]bool)
 
-	for _, match := range matches {
-		if len(match) != 2 {
+	for _, idMatch := range idMatches {
+		if len(idMatch) != 2 {
 			continue
 		}
-		resMatch := strings.TrimSpace(match[1])
-		for _, idmatch := range idMatches {
-			if len(match) != 2 {
+		resIDMatch := strings.TrimSpace(idMatch[1])
+
+		if verifiedIDs[resIDMatch] {
+			continue
+		}
+
+		for _, keyMatch := range keyMatches {
+			if len(keyMatch) != 2 {
+				continue
+			}
+			resKeyMatch := strings.TrimSpace(keyMatch[1])
+
+			if verifiedKeys[resKeyMatch] {
 				continue
 			}
-			resIdMatch := strings.TrimSpace(idmatch[1])
 
 			s1 := detectors.Result{
 				DetectorType: detectorspb.DetectorType_Bulksms,
-				Raw:          []byte(resMatch),
-				RawV2:        []byte(resMatch + resIdMatch),
+				Raw:          []byte(resKeyMatch),
+				RawV2:        []byte(resKeyMatch + resIDMatch),
 			}
```

Comment:
Few suggestions:

- Can remove trim space funcs as there is no space in regex
- Instead of creating `map[string]bool`, we can define `map[string]struct{}` and filter unique values before processing the matches. This will also keep it same as other detectors.

Suggested Code: (This is a suggestion code, you can run it locally and verify and make changes if required)

```
	var uniqueIds = make(map[string]struct{})
	var uniqueKeys = make(map[string]struct{})

	for _, match := range idPat.FindAllStringSubmatch(dataStr, -1) {
		uniqueIds[match[1]] = struct{}{}
	}

	for _, match := range keyPat.FindAllStringSubmatch(dataStr, -1) {
		uniqueKeys[match[1]] = struct{}{}
	}

	for id := range uniqueIds {
		for key := range uniqueKeys {
			s1 := detectors.Result{
				DetectorType: detectorspb.DetectorType_Bulksms,
				Raw:          []byte(key),
				RawV2:        []byte(key + id),
			}

			if verify {
				req, err := http.NewRequestWithContext(ctx, "GET", "https://api.bulksms.com/v1/messages", nil)
				if err != nil {
					continue
				}
				req.SetBasicAuth(id, key)
				res, err := client.Do(req)
				if err == nil {
					defer func() {
						_, _ = io.Copy(io.Discard, res.Body)
						_ = res.Body.Close()
					}()

					if res.StatusCode == http.StatusOK {
						s1.Verified = true
						results = append(results, s1)
						// move to next key, by skipping remaining id's
						break
					}
				} else {
					s1.SetVerificationError(err, key)
				}
			}

			results = append(results, s1)
		}
	}
```

Reply:
Thanks @kashifkhan0771 , this method is better, I agree. I have made the changes and tested it too, please check.